### PR TITLE
[BEAM-7923] Screendiff Integration Tests

### PR DESCRIPTION
**Pull request by KevinGG**
1. Added a screendiff integration test infrastructure: https://docs.google.com/document/d/1tk2P0Cu9hMjbtDWfIuQEFj07q20iqasqLPVtYVpxw3w/edit?usp=sharing
2. Added the first test notebook: init_square_cube.ipynb with golden
   screenshot 7a35f487b2a5f3a9b9852a8659eeb4bd.png and its test: init_square_cube_test.py.
3. To add a new test, simply follow the above example. To generate a new
   golden, run tests with `nosetests apache_beam/runners/interactive/testing/integration/tests/xxx_test.py --with-save-baseline`.
4. Added new dependency group [interactive_test].
5. The test is not part of the Jenkins pre-commit process yet. To run it
   locally, make sure both [interactive] and [interactive_test]
   dependencies are installed, and you have chrome installed.
6. To enable such integration tests on Jenkins, chrome driver plugin
   should be installed.
7. The BaseTestCase provided in the screen_diff module supports running
   tests for all notebooks under a directory, a single test for a golden
   or a single test for a notebook.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by KevinGG** [comment]
yapf formatted.
lint passed locally.

R: @aaltay 
R: @davidyan74 
R: @rohdesamuel 

PTAL, thx!

File: sdks/python/apache_beam/runners/interactive/testing/integration/screen_diff.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  from needle.driver import NeedleChrome
+  from selenium.webdriver.chrome.options import Options
+  _interactive_integration_ready = (
+      notebook_executor._interactive_integration_ready)
+except ImportError:
+  _interactive_integration_ready = False
+
+
+class ScreenDiffIntegrationTestEnvironment(object):
+  """A test environment to conduct screen diff integration tests for notebooks.
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/testing/integration/screen_diff.py:46]: Could we compare produced html outputs? Do we need to compare that pixels are equivalent, if instead we can compare the html outputs?

File: sdks/python/apache_beam/runners/interactive/testing/integration/tests/init_square_cube_test.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+from apache_beam.runners.interactive.testing.integration.screen_diff import BaseTestCase
+
+
+@unittest.skipIf(
+    BaseTestCase.should_skip(),
+    '[interactive] and [interactive_test] dependency are both required.')
+class InitSquareCubeTest(BaseTestCase):
+  def test_init_square_cube_notebook(self):
+    self.assert_notebook('init_square_cube')
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/testing/integration/tests/init_square_cube_test.py:33]: High level question, do we need to compare that pixels did not change with chnages? What benefit do we get from it?

File: sdks/python/setup.py [hunk 3]
```diff
@@ -222,6 +222,17 @@ def get_version():
     'ipython>=5.8.0,<8',
     'timeloop>=1.0.2,<2',
 ]
+
+INTERACTIVE_BEAM_TEST = [
```
Comment (aaltay) [inline sdks/python/setup.py:226]: This is a quite a large addition for the value we are getting.

And,
- should this be integrated with tox?
- jenkins?
- which platforms does this test work in?

File: sdks/python/apache_beam/runners/interactive/testing/integration/notebook_executor.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    self._output_html_dir = os.path.join(self._dir, 'output')
+    self.cleanup()
+    self._output_html_paths = {}
+    self._notebook_path_to_execution_id = {}
+
+  def cleanup(self):
+    """Cleans up the output folder."""
+    _cleanup(self._output_html_dir)
+
+  def execute(self):
```
[the remark this answers is not in the document]
Comment (aaltay) [reply sdks/python/apache_beam/runners/interactive/testing/integration/notebook_executor.py:77]: OK this is fair.

File: sdks/python/setup.py (hunk 3 above)
Comment (aaltay) [reply sdks/python/setup.py:227]: Fair. Please add this information to the wiki. (https://cwiki.apache.org/confluence/display/BEAM/Python+Tips or a new notebook tips page)

File: sdks/python/setup.py [hunk 5]
```diff
@@ -220,8 +220,21 @@ def get_version():
 INTERACTIVE_BEAM = [
     'facets-overview>=1.0.0,<2',
     'ipython>=5.8.0,<8',
+    'ipykernel>=5.2.0,<6',
```
Comment (aaltay) [inline sdks/python/setup.py:223]: Why is this added here and not to the test list?

File: sdks/python/setup.py (hunk 3 above)
Comment (KevinGG) [reply sdks/python/setup.py:227]: Will update the readme in another PR once this is submitted

File: sdks/python/setup.py (hunk 5 above)
Comment (KevinGG) [reply sdks/python/setup.py:223]: It's also part of the Jupyter notebook setup dependency, see the [readme](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/runners/interactive).
We didn't put `jupyterlab` nor `ipykernel` in `[interactive]` because we wanted to keep the dependency minimal sized. Since now we may need `ipykernel` in the virtual env setup,  add it to `[interactive]` because it's where it should go.

Comment (aaltay) [reply sdks/python/setup.py:223]: Got it. This is fine. I got confused because I could not tell its relation to this PR. Please break different changes to different PRs.

Comment (KevinGG) [reply sdks/python/setup.py:223]: It's actually related, because the nbconvert module requires an ipykernel to work.
So this dependency is needed by both `[interactive]` and `[interactive_test]`.

**Comment by aaltay** [comment]
retest this please

**Comment by KevinGG** [comment]
A Jira ticket was created 3 hours ago about the error https://issues.apache.org/jira/browse/BEAM-9736

Global name `from_container_image` is not defined.

**Comment by aaltay** [comment]
> A Jira ticket was created 3 hours ago about the error https://issues.apache.org/jira/browse/BEAM-9736
> 
> Global name `from_container_image` is not defined.

This is marked as fixed now.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please
